apps/crn-server/src/data-providers/contentful/teams.data-provider.ts

```diff
@@ -0,0 +1,241 @@
+  NonNullable<FetchTeamsQuery['teamsCollection']>['items'][number]
+>;
+
+type TeamFilter = {
+  active?: boolean;
+};
+
+export type FetchTeamsOptions = FetchOptions<TeamFilter>;
+export class TeamContentfulDataProvider implements TeamDataProvider {
+  constructor(
+    private contentfulClient: GraphQLClient,
+    private getRestClient: () => Promise<Environment>,
+  ) {}
+
+  async fetch(options: FetchTeamsOptions): Promise<ListTeamDataObject> {
+    const { take = 8, skip = 0, search, filter } = options;
+
+    const searchTerms = (search || '').split(' ').filter(Boolean);
+    const searchQuery = searchTerms.length
+      ? {
+          OR: [
+            ...searchTerms.map((term) => ({
+              displayName_contains: term,
+            })),
+            ...searchTerms.map((term) => ({
+              projectTitle_contains: term,
+            })),
+            {
+              expertiseAndResourceTags_contains_some: searchTerms,
+            },
+          ],
+        }
+      : {};
+
+    const activeQuery =
+      typeof filter?.active === 'boolean'
+        ? { inactiveSince_exists: !filter?.active }
+        : {};
+
+    const { teamsCollection } = await this.contentfulClient.request<
+      FetchTeamsQuery,
+      FetchTeamsQueryVariables
+    >(FETCH_TEAMS, {
+      limit: take,
+      skip,
+      order: [TeamsOrder.DisplayNameAsc],
+      where: {
+        ...searchQuery,
+        ...activeQuery,
+      },
+    });
+
+    if (!teamsCollection?.items) {
+      return {
+        total: 0,
+        items: [],
+      };
+    }
+
+    return {
+      total: teamsCollection?.total,
+      items: teamsCollection?.items
+        .filter((x): x is TeamItem => x !== null)
+        .map(parseContentfulGraphQlTeams),
+    };
+  }
+
+  async fetchById(id: string): Promise<TeamDataObject | null> {
+    const { teams } = await this.contentfulClient.request<
+      FetchTeamByIdQuery,
+      FetchTeamByIdQueryVariables
+    >(FETCH_TEAM_BY_ID, { id });
+
+    if (!teams) {
+      return null;
+    }
+
+    return parseContentfulGraphQlTeams(teams);
+  }
+
+  async update(id: string, update: TeamUpdateDataObject): Promise<void> {
+    const environment = await this.getRestClient();
+
+    const cleanTools = getCleanTools(update.tools);
+
+    const publishedTools = await createAndPublishTools(environment, cleanTools);
+
+    const team = await environment.getEntry(id);
+
+    const teamUpdated = await team.patch([
+      {
+        op: 'replace',
+        path: '/fields/tools/en-US',
+        value: publishedTools.map((tool) => ({
+          sys: {
+            type: 'Link',
+            linkType: 'Entry',
+            id: tool.sys.id,
+          },
+        })),
+      },
+    ]);
+
+    await teamUpdated.publish();
+  }
+
+  async create(input: TeamCreateDataObject): Promise<string> {
+    const environment = await this.getRestClient();
+
+    const { tools, ...plainInput } = input;
+    let toolsInput = {};
+
+    if (tools) {
```

Comment:
Is it possible to create the team first then add tools to them. As currently if we create some tools and any fail except the first one, we would have tools in the CMS that don't belong to a team. Creating the team first would mean at least successfully tool creations would be linked to the team.

Reply:
> Creating the team first would mean at least successfully tool creations would be linked to the team.
I didn't understand. Could you explain it again?

Reply:
Let me try explaining with an example

Say we are creating a team with two tools (A,B) and creating tool B fails the whole createTeam request will fail. So if we retry creating the team and it succeeds the CMS would have the tools (A,A,B) where one A doesn't belong to a team.